From 06c5d535332898944d627b6e114d1f39ae3e6c86 Mon Sep 17 00:00:00 2001
From: Mikhail Naganov <mnaganov@google.com>
Date: Fri, 14 Dec 2018 10:48:17 -0800
Subject: [PATCH 48/80] audioserver: Use '_exit' instead of 'exit' in
 HalDeathHandler

Using 'exit' from an RPC threadpool thread is not safe, as 'exit'
runs atexit handlers that destroy global objects. This can interfere
with code still running on other threads.

'_exit' does not run atexit handlers, just terminates the process.

Bug: 116665972
Test: kill android.hardware.audio@2.0-service, check logcat
Change-Id: I5391a659e359e0ca5bba580f1c51dea5df3ea562
(cherry picked from commit 5beb4810060cda2afea1c89f98caf5c337763c0a)
---
 media/libaudiohal/HalDeathHandlerHidl.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/media/libaudiohal/HalDeathHandlerHidl.cpp b/media/libaudiohal/HalDeathHandlerHidl.cpp
index 1e3ab5846..6e3352309 100644
--- a/media/libaudiohal/HalDeathHandlerHidl.cpp
+++ b/media/libaudiohal/HalDeathHandlerHidl.cpp
@@ -54,7 +54,7 @@ void HalDeathHandler::serviceDied(uint64_t /*cookie*/, const wp<IBase>& /*who*/)
         handler.second();
     }
     ALOGE("HAL server crashed, audio server is restarting");
-    exit(1);
+    _exit(1);  // Avoid calling atexit handlers, as this code runs on a thread from RPC threadpool.
 }
 
 } // namespace android
-- 
2.17.1

